### PR TITLE
Added setting UsePolymorphicSerialization

### LINKED ISSUE #459: Is it possible to replace the liquid files of NSwag?
NSwag has a bug in their generated code (see [here](https://github.com/RicoSuter/NJsonSchema/issues/1728) and see the [fix](https://github.com/RicoSuter/NJsonSchema/pull/1729))

It unfortunately seems the author is not very actively closing PRs.
Meanwhile I'm looking for options to 'fix' here, if possible?

We could alter the generated code and commit it, but in out CI the code gets generated again.
Also after each update, we would have to alter the code again, which can become cumbersome very fast.

It would be a good alternative if Refitter could 'fix' this issue in the meantime.

### PR DESCRIPTION
**Pull request by ryanheath**
Fixes #459 and [Unknown subtype causes Stackoverflow exception at deserialization](https://github.com/RicoSuter/NJsonSchema/issues/1728)

Loosely based on [Add support for STJ-native C# code gen](https://github.com/RicoSuter/NJsonSchema/pull/1675)

When UseSystemTextJsonPolymorphicSerialization is set to true
a CustomTemplateFactory is used that overrides the NSwag DefaultTemplateFactory behavior.
The STJ polymorphic attributes are rendered instead of NSwag own polymorphic attributes.

### CODE REVIEW

**Comment by christianhelle** [comment]
@ryanheath Thanks for taking the time to investigate and implement this. I'll start reviewing this immediately

**Comment by christianhelle** [comment]
@all-contributors please add @ryanheath for code

File: src/Refitter.Core/Settings/RefitGeneratorSettings.cs [hunk 1]
```diff
@@ -216,4 +216,10 @@ public class RefitGeneratorSettings
...
     /// </summary>
     public bool GenerateMultipleFiles { get; set; }
-}
\ No newline at end of file
+
+    /// <summary>
+    /// Set to <c>true</c> to use System.Text.Json polymorphic serialization. Default is <c>false</c>
+    /// Gets a value indicating whether to use System.Text.Json polymorphic serialization
+    /// </summary>
+    public bool UseSystemTextJsonPolymorphicSerialization { get; set; }
```
Comment (christianhelle) [inline src/Refitter.Core/Settings/RefitGeneratorSettings.cs]: @ryanheath The name is a bit long, can we rename this to `UsePolymorphicSerialization` ? Refitter explicitly configures NSwag to use System.Text.Json so I was thinking that there is no need to repeat it

File: src/Refitter.Tests/Examples/UseSystemTextJsonPolymorphicSerializationTests.cs [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+    [Fact]
+    public async Task Can_Generate_Code()
+    {
+        string generateCode = await GenerateCode();
+        generateCode.Should().NotBeNullOrWhiteSpace();
+    }
+
+    [Fact]
+    public async Task Use_SystemTextJson_Polymorphic_Serialization()
```
Comment (christianhelle) [inline src/Refitter.Tests/Examples/UseSystemTextJsonPolymorphicSerializationTests.cs]: @ryanheath This test seems to fail. Can you please fix it?

File: src/Refitter/Settings.cs [hunk 3]
```diff
@@ -210,4 +210,9 @@ The NSwag IOperationNameGenerator implementation to use.
     [CommandOption("--use-dynamic-querystring-parameters")]
     [DefaultValue(false)]
     public bool UseDynamicQuerystringParameters { get; set; }
-}
\ No newline at end of file
+
+    [Description("Set to <c>true</c> to use System.Text.Json polymorphic serialization.")]
+    [CommandOption("--use-system-text-json-polymorphic-serialization")]
```
Comment (christianhelle) [inline src/Refitter/Settings.cs]: @ryanheath Can we rename the command option to be `--use-polymorphic-serialization` ?

File: src/Refitter.Core/Refitter.Core.csproj [hunk 4]
```diff
@@ -23,4 +23,8 @@
     </AssemblyAttribute>
   </ItemGroup>
 
+  <ItemGroup>
```
Comment (christianhelle) [inline src/Refitter.Core/Refitter.Core.csproj:26]: @ryanheath We must include the liquid templates as embedded resources in the source generator project `src/Refitter.SourceGenerator/Refitter.SourceGenerator.csproj` as well

**Review by christianhelle — CHANGES_REQUESTED** [review]
@ryanheath I just have a few minor suggestions, but all-in-all this looks great!

File: src/Refitter.Tests/Examples/UseSystemTextJsonPolymorphicSerializationTests.cs (hunk 2 above)
Comment (ryanheath) [reply src/Refitter.Tests/Examples/UseSystemTextJsonPolymorphicSerializationTests.cs]: It only seems to fail when run in parallel, I'm not sure why.

File: src/Refitter.Core/Refitter.Core.csproj (hunk 4 above)
Comment (ryanheath) [reply src/Refitter.Core/Refitter.Core.csproj:26]: Hmm, is Refitter.SourceGenerator.csproj not referencing Refitter.Core.csproj?
The way the resources are loaded, is that they are bound to assemblies names.

Comment (christianhelle) [reply src/Refitter.Core/Refitter.Core.csproj:26]: @ryanheath No, files are included as links instead. It was a bit tricky to get the source generator up and running and we kind of got stuck where we don't add files directly to the source generator project but only work on the core project. 

I should probably re-visit the source generator project setup again. You can skip the change request, I'll take care of it once we merge this in

File: src/Refitter.Tests/Examples/UseSystemTextJsonPolymorphicSerializationTests.cs (hunk 2 above)
Comment (christianhelle) [reply src/Refitter.Tests/Examples/UseSystemTextJsonPolymorphicSerializationTests.cs]: @ryanheath I pulled your branch and tried it out locally. You're right, it only fails when run in parallel. The tests are correct and the structure is fine. I can't immediately see what could be wrong but the generated code is quite different...

When run individually, it generates 244 lines of code, when run in parallel it generates 375 lines of code 🤯

File: src/Refitter.Core/Settings/RefitGeneratorSettings.cs (hunk 1 above)
Comment (ryanheath) [reply src/Refitter.Core/Settings/RefitGeneratorSettings.cs]: Changed

File: src/Refitter.Tests/Examples/UseSystemTextJsonPolymorphicSerializationTests.cs (hunk 2 above)
Comment (ryanheath) [reply src/Refitter.Tests/Examples/UseSystemTextJsonPolymorphicSerializationTests.cs]: There is a static concurrentdictionary that holds the templates. I'll try to see how to fix this problem.
We might need to override a lot more than I really wanted ...

File: src/Refitter/Settings.cs (hunk 3 above)
Comment (ryanheath) [reply src/Refitter/Settings.cs]: Changed

File: src/Refitter.Core/Refitter.Core.csproj (hunk 4 above)
Comment (ryanheath) [reply src/Refitter.Core/Refitter.Core.csproj:26]: I have made the change request, it was needed anyway to get the source generator work the same as core.

**Comment by ryanheath** [comment]
The tests are still failing because of a static template dictionary deep down in the nwag or nschema sources.
I'll see want can be done.

**Comment by ryanheath** [comment]
I have added a lot more override and custom implementation of NSwag/NSchema generators and models.
NSwag/NSchema is not extensible friendly regarding this point, or I must have overlooked an easy way to access settings in the template files ... 😅

**Review by christianhelle — APPROVED** [review]
Great work @ryanheath

**Comment by christianhelle** [comment]
> I have added a lot more override and custom implementation of NSwag/NSchema generators and models. NSwag/NSchema is not extensible friendly regarding this point, or I must have overlooked an easy way to access settings in the template files ... 😅

Yeah, I agree. It looks needlessly complicated and harder than it should be.

Let's merge this in

**Comment by ryanheath** [comment]
Thank you @christianhelle!

**Comment by ryanheath** [comment]
@christianhelle what else should we do to get a new (pre) release nuget?

**Comment by christianhelle** [comment]
> @christianhelle what else should we do to get a new (pre) release nuget?

I can create a pre-release now as it is so that you can use it

**Comment by christianhelle** [comment]
@ryanheath I was planning to update the docs tonight but then I can just do another pre-release with the new docs

**Pull request merged**